### PR TITLE
Export custom statuses in api export

### LINKED ISSUE #1942: Registrants export API does not contain custom statuses
> I tried to access registrants data for an upcoming conference we are
> organizing 
> I was able to extract all registration data as described on http://indico.readthedocs.org/en/latest/http_api/exporters/registration/#full
> But I could not get access to the custom statuses which can be defined
> i.e. at the registration setup in the managament interface

Probably makes sense to include this information...

### PR DESCRIPTION
**Pull request by mic4ael**
fixes #1942

### CODE REVIEW

File: indico/MaKaC/registration.py [hunk 1]
```diff
@@ -585,6 +585,13 @@ def getETicket(self):
             self._eTicket = eticket.ETicket()
             return self._eTicket
 
+    def hasStatusDefined(self, status_name):
+        for status in self._statuses.values():
```
Comment (ThiefMaster) [inline indico/MaKaC/registration.py]: return any(status.getCaption() == status_name for status in self._statuses.itervalues())

File: indico/MaKaC/registration.py [hunk 2]
```diff
@@ -4822,6 +4829,10 @@ def getStatuses(self):
     def getStatusesList(self):
         return self.getStatuses().values()
 
+    def getStatusesWithValues(self):
+        return {status.getCaption(): status.getStatusValue().getCaption()
+                for status in self._statuses.values()}
```
Comment (ThiefMaster) [inline indico/MaKaC/registration.py]: itervalues

File: indico/MaKaC/registration.py [hunk 3]
```diff
@@ -585,6 +585,13 @@ def getETicket(self):
             self._eTicket = eticket.ETicket()
             return self._eTicket
 
+    def hasStatusDefined(self, status_name):
```
Comment (ThiefMaster) [inline indico/MaKaC/registration.py]: camelcase = big-no

File: indico/MaKaC/webinterface/rh/registrationFormModif.py [hunk 4]
```diff
@@ -160,7 +160,8 @@ def __init__(self, rh, conf, params):
         self._params=params
 
     def perform( self ):
-        if self._params.get("caption","").strip() !="":
+        status_caption = self._params.get("caption", "").strip()
```
Comment (ThiefMaster) [inline indico/MaKaC/webinterface/rh/registrationFormModif.py]: single quotes please :)

File: indico/MaKaC/webinterface/rh/registrationFormModif.py [hunk 5]
```diff
@@ -228,8 +229,8 @@ def map(self, params):
         if params.get("addvalue","").strip()!="":
             sv=StatusValue(self)
             cap=params.get("newvalue")
-            if cap.strip()!="":
-                sv.setCaption(params.get("newvalue"))
+            if cap.strip() != "" and cap not in map(lambda val: val.getCaption(), self._values):
```
Comment (ThiefMaster) [inline indico/MaKaC/webinterface/rh/registrationFormModif.py]: `!= ''` can be implied

File: indico/core/fossils/registration.py [hunk 6]
```diff
@@ -812,3 +812,12 @@ def getMiscellaneousGroupList(self):
         Get Miscellaneous group list
         """
     getMiscellaneousGroupList.result = IRegFormMiscellaneousInfoGroupFullFossil
+
+    def getStatuses(self):
+        """
```
Comment (ThiefMaster) [inline indico/core/fossils/registration.py]: no need for a docstring here. it's a legacy thing and just defines an interface anyway

File: indico/MaKaC/registration.py (hunk 3 above)
Comment (mic4ael) [inline indico/MaKaC/registration.py]: i do know, however i left it that way to conform to the rest of the code.

File: indico/MaKaC/webinterface/rh/registrationFormModif.py [hunk 7]
```diff
@@ -228,8 +229,8 @@ def map(self, params):
         if params.get("addvalue","").strip()!="":
             sv=StatusValue(self)
             cap=params.get("newvalue")
-            if cap.strip()!="":
-                sv.setCaption(params.get("newvalue"))
+            if cap.strip() and cap not in map(lambda val: val.getCaption(), self._values):
```
Comment (ThiefMaster) [inline indico/MaKaC/webinterface/rh/registrationFormModif.py]: once again, don't pass a lambda to map

`not any(cap == val.getCaption() for val in self._values)` is much more efficient here. another option would be `cap not in {val.getCaption() for val in self._values}` but why build a set when you don't really need it

File: indico/MaKaC/registration.py [hunk 8]
```diff
@@ -4822,6 +4825,10 @@ def getStatuses(self):
     def getStatusesList(self):
         return self.getStatuses().values()
 
+    def getStatusesWithValues(self):
```
Comment (ThiefMaster) [inline indico/MaKaC/registration.py]: camelcase

File: indico/MaKaC/registration.py [hunk 9]
```diff
@@ -4822,6 +4825,10 @@ def getStatuses(self):
...
+        return {status.getCaption(): status.getStatusValue().getCaption()
```
Comment (ThiefMaster) [inline indico/MaKaC/registration.py]: can there be two statuses with the same caption?

Comment (mic4ael) [inline indico/MaKaC/registration.py]: Nope, I have fixed that some time ago.

**Comment by ThiefMaster** [comment]
Obsolete, but we should keep this functionality in mind for 2.0 or whenever we cleanup/stabilize APIs.

**Pull request closed without merging**